New .NET diagnostics NuGet packages content. (#37455)
dotnet/docs: docs/core/diagnostics/diagnostic-libraries.md

---
title: Diagnostic libraries
description: An overview of the available diagnostic libraries in .NET.
ms.date: 10/16/2023
ms.topic: overview
---

# Diagnostic libraries

Diagnostic libraries are tools to collect and analyze diagnostic information about your .NET applications. From summarizing exceptions, determining resource utilization and monitoring, to exposing health checks, and providing probes for Kubernetes. The diagnostic packages provide a rich set of functionality that can be used to instrument your apps.

> The following NuGet packages are available:

Reviewer: There are other diagnostic libraries supplied in .NET, those listed here happen to be the new ones. I think the title and description of this article are thus misleading, and perhaps this whole article should be split.

There is already content in the .NET docs about health checks, so I think the health-related stuff should move there. https://learn.microsoft.com/en-us/aspnet/core/host-and-deploy/health-checks?view=aspnetcore-7.0

The health probe topic is about reporting the process health to the environment, so that the environment can stop routing traffic to the process when it is unhealthy. This should probably go into ASP.NET docs.

The exception summarization is used in telemetry (mainly for metrics). So it should go into a telemetry-related area.

Author: > There are other diagnostic libraries supplied in .NET, those listed here happen to be the new ones. I think the title and description of this article are thus misleading, and I think this whole article should be split since it combines things together which don't belong together.

I'm not opposed to the splitting of content, could you please help direct me as to where you think each bit of content fits within the .NET docs? Here is a link to the current .NET Diagnostics content: https://learn.microsoft.com/dotnet/core/diagnostics, as I believe it should be added somewhere within this subset of content.

> There is already content in the .NET docs about health checks, so I think the health-related stuff should move there. https://learn.microsoft.com/en-us/aspnet/core/host-and-deploy/health-checks?view=aspnetcore-7.0

The link you provided is the ASP.NET Core docs, not .NET docs. We are workload agnostic. We can link to that content for additional reading as it relates to ASP.NET Core (or web-specific workloads). Also, their content touches on the primitives, but exposes this functionality through middleware — a package that sits atop these two health check packages.

> The health probe topic is about reporting the process health to the environment, so that the environment can stop routing traffic to the process when it is unhealthy. This should probably go into ASP.NET docs.

It seems like this library isn't ready yet anyway, perhaps we exclude it for now? Also, it's specific to Kubernetes, no? I don't think of that being synonymous with ASP.NET docs, as that is workload agnostic.

Author: I split the content and attempted to put it where I thought it made the most sense.

> Available as a set of .NET Extensions, various diagnostic libraries provide a set of APIs and tools that enable you to collect and analyze diagnostic information about your .NET applications. From summarizing exceptions, to determining resource utilization and monitoring, to exposing health checks, and providing probes for Kubernetes, the diagnostic packages provide a rich set of functionality that can be used to instrument your apps.

Reviewer 2: ```suggestion
Diagnostic libraries are tools to collect and analyze diagnostic information about your .NET applications. From summarizing exceptions, determining resource utilization and monitoring, to exposing health checks, and providing probes for Kubernetes. The diagnostic packages provide a rich set of functionality that can be used to instrument your apps.
```

IMHO shorter without losing any information.

| NuGet package | Description |
|--|--|

> | [📦 Microsoft.Extensions.Diagnostics.ExceptionSummarization](https://www.nuget.org/packages/Microsoft.Extensions.Diagnostics.ExceptionSummarization) | Provides a mechanism for retrieving exception summary information. |

Reviewer 2: I would drop 'Provides a' || 'Provides' prefixes to make it shorter and less repetetive.

> | [📦 Microsoft.Extensions.Diagnostics.ExceptionSummarization](https://www.nuget.org/packages/Microsoft.Extensions.Diagnostics.ExceptionSummarization) | Provides a mechanism for retrieving exception summary information. |
> | [📦 Microsoft.Extensions.Diagnostics.ResourceMonitoring](https://www.nuget.org/packages/Microsoft.Extensions.Diagnostics.ResourceMonitoring) | Provides resource monitoring that measures processor and memory usage. |
> | [📦 Microsoft.Extensions.Diagnostics.HealthChecks.Common](https://www.nuget.org/packages/Microsoft.Extensions.Diagnostics.HealthChecks.Common) | Provides a standard health check implementation, as well as host application lifetime hooks. |
> | [📦 Microsoft.Extensions.Diagnostics.HealthChecks.ResourceUtilization](https://www.nuget.org/packages/Microsoft.Extensions.Diagnostics.HealthChecks.ResourceUtilization) | Provides resource utilization health checks. |
> | [📦 Microsoft.Extensions.Diagnostics.Probes](https://www.nuget.org/packages/Microsoft.Extensions.Diagnostics.Probes) | Provides support for Kubernetes environmental probes. |

Author: ```suggestion
| [📦 Microsoft.Extensions.Diagnostics.ExceptionSummarization](https://www.nuget.org/packages/Microsoft.Extensions.Diagnostics.ExceptionSummarization) | Mechanisms for retrieving exception summary information. |
| [📦 Microsoft.Extensions.Diagnostics.ResourceMonitoring](https://www.nuget.org/packages/Microsoft.Extensions.Diagnostics.ResourceMonitoring) | Resource monitoring that measures processor and memory usage. |
| [📦 Microsoft.Extensions.Diagnostics.HealthChecks.Common](https://www.nuget.org/packages/Microsoft.Extensions.Diagnostics.HealthChecks.Common) | Standard health check implementation, as well as host application lifetime hooks. |
| [📦 Microsoft.Extensions.Diagnostics.HealthChecks.ResourceUtilization](https://www.nuget.org/packages/Microsoft.Extensions.Diagnostics.HealthChecks.ResourceUtilization) | Resource utilization health checks. |
| [📦 Microsoft.Extensions.Diagnostics.Probes](https://www.nuget.org/packages/Microsoft.Extensions.Diagnostics.Probes) | Support for Kubernetes environmental probes. |
```

## Exception summarization

> When you're trying to generate meaningful diagnostic messages for exceptions, maintaining the inclusion of pertinent information can pose a challenge. The standard exception message often lacks critical details that accompany the exception, while invoking the <xref:System.Exception.ToString%2A?displayProperty=nameWithType> method yields an excess of state information.

Reviewer 2: ```maintaining the inclusion of pertinent information can pose a challenge```

as non native English reader I am struggling to understand what does it mean. Could we write it using simpler wording?

like (if I understood correctly) - `generate meaningful diagnostic messages for exceptions without leaking sensitive data can be challenging`

> When you're trying to generate meaningful diagnostic messages for exceptions, maintaining the inclusion of pertinent information can pose a challenge. The standard exception message often lacks critical details that accompany the exception, while invoking the <xref:System.Exception.ToString%2A?displayProperty=nameWithType> method yields an excess of state information.

Author: ```suggestion
When you're trying to generate meaningful diagnostic messages for exceptions without leaking sensitive data it's challenging. The standard exception message often lacks critical details that accompany the exception, while invoking the <xref:System.Exception.ToString%2A?displayProperty=nameWithType> method yields an excess of state information.
```

> The <xref:Microsoft.Extensions.Diagnostics.ExceptionSummarization.IExceptionSummarizer> interface offers methods for extracting crucial details from recognized exception types, thereby furnishing a singular `string` that serves as the foundation for crafting top-quality diagnostic messages.

Reviewer 2: `thereby furnishing a singular 'string' that serves as the foundation for crafting top-quality`

Does it mean transforming to a string?

> The <xref:Microsoft.Extensions.Diagnostics.ExceptionSummarization.IExceptionSummarizer.Summarize%2A?displayProperty=nameWithType> method systematically traverses the roster of registered summarizers until it identifies a summarizer capable of handling the specific exception type. In the event that no summarizer is capable of recognizing the exception type, a meaningful default exception summary is provided instead.

Reviewer 2: ```suggestion
The <xref:Microsoft.Extensions.Diagnostics.ExceptionSummarization.IExceptionSummarizer.Summarize%2A?displayProperty=nameWithType> method traverses registered summarizers until it identifies one capable of handling the specific exception type. In the event that no summarizer is capable of recognizing the exception type, fallback summary is provided instead.
```

IMHO shorter

The following example demonstrates how to use the `IExceptionSummarizer` interface to retrieve a summary of an exception.

:::code source="snippets/exception-summary/Program.cs":::

The preceding code:

- Instantiates a new <xref:Microsoft.Extensions.DependencyInjection.ServiceCollection> instance, chaining a call to the <xref:Microsoft.Extensions.Diagnostics.ExceptionSummarization.ExceptionSummarizationExtensions.AddExceptionSummarizer%2A> extension method.
  - The `AddExceptionSummarizer` extension method accepts a delegate that is used to configure the `ExceptionSummarizerBuilder` instance.
  - The `builder` is used to add the HTTP provider, which handles exceptions of type:
    - <xref:System.OperationCanceledException>
    - <xref:System.Threading.Tasks.TaskCanceledException>
    - <xref:System.Net.Sockets.SocketException>
    - <xref:System.Net.WebException>
- Builds a new `ServiceProvider` instance from the `ServiceCollection` instance.
- Gets an instance of the `IExceptionSummarizer` interface from the `ServiceProvider` instance.
- Iterates over a collection of exceptions, calling the `Summarize` method on each exception and displaying the result.

> [!WARNING]

> > The primary focus in the design of all exception summarization implementations is to provide diagnostic convenience to trusted users, rather than prioritizing the protection of personally identifiable information (PII). The <xref:Microsoft.Extensions.Diagnostics.ExceptionSummarization.ExceptionSummary.Description?displayProperty=nameWithType> doesn't contain sensitive information, but the <xref:Microsoft.Extensions.Diagnostics.ExceptionSummarization.ExceptionSummary.AdditionalDetails?displayProperty=nameWithType> might contain sensitive information depending on the implementation.

Reviewer: I don't know whay "provide diagnostic convenience to trusted users" means.

The goal of summarization is twofold:

* To eliminate privacy-sensitive information from exception state such that some meaningful exception information can be added to logs.
* To reduce the cardinality associated with exception state such that exceptions can be reliably counted in metrics. This matters since metric dimensions have limited cardinality.

Author: This language was adopted from this response here, https://github.com/dotnet/extensions/issues/4403#issuecomment-1718402588, but happy to update it.

## Resource monitoring

Resource monitoring involves the continuous measurement of resource utilization over a specified period. The [Microsoft.Extensions.Diagnostics.ResourceMonitoring](https://www.nuget.org/packages/Microsoft.Extensions.Diagnostics.ResourceMonitoring) NuGet package offers a collection of APIs tailored for monitoring the resource utilization of your .NET applications.

> The <xref:Microsoft.Extensions.Diagnostics.ResourceMonitoring.IResourceMonitor> interface furnishes methods for retrieving real-time information concerning process resource utilization. This interface supports the retrieval of data related to CPU and memory usage and is currently compatible with both Windows and Linux platforms.

Reviewer 2: I think it is worth to mention, that this information is published to OTel by default, so one does not need to publish it on themselves.

- <xref:Microsoft.Extensions.Diagnostics.ResourceMonitoring.ResourceUtilization.MemoryUsedPercentage?displayProperty=nameWithType>: Memory usage as a percentage.
- <xref:Microsoft.Extensions.Diagnostics.ResourceMonitoring.ResourceUtilization.MemoryUsedInBytes?displayProperty=nameWithType>: Memory usage in bytes.
- <xref:Microsoft.Extensions.Diagnostics.ResourceMonitoring.ResourceUtilization.SystemResources?displayProperty=nameWithType>: System resources.
  - <xref:Microsoft.Extensions.Diagnostics.ResourceMonitoring.SystemResources.GuaranteedMemoryInBytes?displayProperty=nameWithType>: Guaranteed memory in bytes.
  - <xref:Microsoft.Extensions.Diagnostics.ResourceMonitoring.SystemResources.MaximumMemoryInBytes?displayProperty=nameWithType>: Maximum memory in bytes.
  - <xref:Microsoft.Extensions.Diagnostics.ResourceMonitoring.SystemResources.GuaranteedCpuUnits?displayProperty=nameWithType>: Guaranteed CPU in units.
  - <xref:Microsoft.Extensions.Diagnostics.ResourceMonitoring.SystemResources.MaximumCpuUnits?displayProperty=nameWithType>: Maximum CPU in units.

> Extending this example, you can leverage [Spectre.Console](https://www.nuget.org/packages/Spectre.Console), a well-regarded .NET library designed to simplify the development of visually appealing, cross-platform console applications. With Spectre, you'll be able to present resource utilization data in a tabular format. The following code illustrates the usage of the `IResourceMonitor` interface to access details regarding the CPU and memory usage of the current process, then presenting this data in a table:

Reviewer 2: `
Extending this example, you can leverage [Spectre.Console](https://www.nuget.org/packages/Spectre.Console), a well-regarded .NET library designed to simplify the development of visually appealing, cross-platform console applications. With Spectre, you'll be able to present resource utilization data in a tabular format.
`

I would remove it, I think it is not related to the content of this document. I think we should not overload  a reader with too much information. We can always create a doc about how to create a tables in .NET.

The following heath check status results are possible:

- <xref:Microsoft.Extensions.Diagnostics.HealthChecks.HealthStatus.Healthy?displayProperty=nameWithType>
- <xref:Microsoft.Extensions.Diagnostics.HealthChecks.HealthStatus.Degraded?displayProperty=nameWithType>
- <xref:Microsoft.Extensions.Diagnostics.HealthChecks.HealthStatus.Unhealthy?displayProperty=nameWithType>

### Application lifetime health checks

To perform health checks on the application lifetime events of <xref:Microsoft.Extensions.Hosting.IHostApplicationLifetime>, use the <xref:Microsoft.Extensions.Diagnostics.HealthChecks.CommonHealthChecksExtensions.AddApplicationLifecycleHealthCheck%2A> extension method available in the [Microsoft.Extensions.Diagnostics.HealthChecks.Common](https://www.nuget.org/packages/Microsoft.Extensions.Diagnostics.HealthChecks.Common) NuGet package.

> This provider will indicate that the application is healthy only when it is fully active. Until the lifetime object indicates the application has started, the provider will report the application as not healthy. When the application starts shutting down, the provider will report the application as unhealthy.

Reviewer 2: ```suggestion
This provider will signal that the application is healthy when lifetime object will indicate the application has started. When the application starts shutting down, the provider will report the application as unhealthy.
```

I've seen a repetition of 'indicates' word twice, and during the process of replacing it, I did more changes that shortens the sentence without losing information.

- Creates a new <xref:Microsoft.Extensions.Hosting.HostApplicationBuilder> instance.
- Adds resource monitoring by calling <xref:Microsoft.Extensions.Diagnostics.ResourceMonitoring.ResourceMonitoringExtensions.AddResourceMonitoring%2A>.
- Adds a health check for resource utilization by chaining a call from the <xref:Microsoft.Extensions.DependencyInjection.HealthCheckServiceCollectionExtensions.AddHealthChecks%2A> call to the <xref:Microsoft.Extensions.Diagnostics.HealthChecks.ResourceUtilizationHealthCheckExtensions.AddResourceUtilizationHealthCheck%2A> extension method.
- Builds the `IHost` instance.
- Gets an instance of the <xref:Microsoft.Extensions.Diagnostics.HealthChecks.HealthCheckService> class from the service provider.
- Performs a health check and displays the result.
- Runs the application.

> [!NOTE]
> The `Microsoft.Extensions.Diagnostics.HealthChecks.ResourceUtilization` library assumes that the consumer will register the dependent call to <xref:Microsoft.Extensions.Diagnostics.ResourceMonitoring.ResourceMonitoringExtensions.AddResourceMonitoring%2A>. If you don't register this, when resolving the `HealthCheckService` an exception is thrown.

## Kubernetes probes

> The [Microsoft.Extensions.Diagnostics.Probes](https://www.nuget.org/packages/Microsoft.Extensions.Diagnostics.Probes) NuGet package provides support for Kubernetes probes. This package is currently experimental and is subject to change.

Reviewer 2: I would not mention here that package is experimental. We have the whole infra for informing developer about it, and mentioning it in docs is simple to forget and get out of sync with code.
